external/fv3fit/fv3fit/keras/_models/shared/clip.py

```diff
@@ -0,0 +1,185 @@
+import dataclasses
+import numpy as np
+import tensorflow as tf
+from typing import Mapping, Hashable, Sequence, Optional, Tuple
+from fv3fit._shared.config import SliceConfig, PackerConfig
+from .sequences import XyMultiArraySequence
+
+ClipDims = Mapping[Hashable, Mapping[str, SliceConfig]]
+
+
+@dataclasses.dataclass(frozen=True)
+class ClipConfig(PackerConfig):
+    clip: ClipDims = dataclasses.field(default_factory=dict)
+
+    def single_feature_slice(self, name: str) -> SliceConfig:
+        self._check_only_feature_dim_is_clipped(name)
+        variable_slice_config = list(self.clip[name].values())[0]
+        return variable_slice_config
+
+    def removed_sizes(self, original_1d_layer_size: int, name: str):
+        # Returns the size of the feature dim slices at the start and end of a
+        # 1D layer that are REMOVED by clipping. Used to pad output layer back to their
+        # original size.
+        variable_slice_config = list(self.clip[name].values())[0]
+        start = variable_slice_config.start or 0
+        stop = variable_slice_config.stop or original_1d_layer_size
+        n_removed_from_start = start
+        n_removed_from_end = original_1d_layer_size - stop
+        return (n_removed_from_start, n_removed_from_end)
+
+    def clip_layer_along_feature_dim(self, layer, name):
```

Comment:
These are all public APIs - can you please add type hints and docstrings where they're missing?

nit: can you rename this to `clip_layer_along_last_dim`? Technically speaking, all non-sample dimensions are feature dimensions. We've been using "feature dimension" to mean the last dimension even if some feature dimensions are x/y instead of z, which is OK conversationally but we should be careful not to put too much in the code when there's a clearer alternative.